<?xml version="1.0"?>
<robot xmlns:xacro="http://ros.org/wiki/xacro" name="primitive_chargepal_with_gripper">

  <!-- constants -->
  <xacro:property name="M_PI" value="3.14159"/>

  <!-- Resolve prefix variable (useful for multi robot) -->
  <xacro:include filename="$(find chargepal_description)/urdf/utils/prefix.urdf.xacro" />


  <!-- ############### PLATFORM ################ -->
  <xacro:include filename="$(find chargepal_description)/urdf/mobile_platform/primitive_mobile_platform.urdf.xacro"/>

  <!-- platform arm joint -->
  <joint name="${prefix}platform_ur10e_joint" type="fixed">
    <parent link="${prefix}tower_link" />
    <child link="${prefix}base_link" />
    <origin xyz="0.0 0.0 ${tower_height/2}" rpy="0.0 0.0 ${0.5*M_PI}" />
  </joint>


  <!-- ############### UR10e ARM ################ -->
  <xacro:include filename="$(find ur_description)/urdf/inc/ur10e_macro.xacro"/>
  <!-- Arguments to load UR10e arm macro -->
  <xacro:arg name="joint_limit_params" default="$(find chargepal_description)/config/ur10e/joint_limits.yaml"/>
  <xacro:arg name="physical_params" default="$(find ur_description)/config/ur10e/physical_parameters.yaml"/>
  <xacro:arg name="kinematics_params" default="$(find ur_description)/config/ur10e/default_kinematics.yaml"/>
  <xacro:arg name="visual_params" default="$(find ur_description)/config/ur10e/visual_parameters.yaml"/>
  <!--
    legal values:
      - hardware_interface/PositionJointInterface
      - hardware_interface/VelocityJointInterface
      - hardware_interface/EffortJointInterface

    NOTE: this value must correspond to the controller configured in the
          controller .yaml files in the 'config' directory.
  -->
  <xacro:arg name="transmission_hw_interface" default="hardware_interface/VelocityJointInterface"/>  
  <xacro:arg name="safety_limits" default="false"/>
  <xacro:arg name="safety_pos_margin" default="0.15"/>
  <xacro:arg name="safety_k_position" default="20"/>

  <xacro:ur10e_robot
    prefix="${prefix}"
    joint_limits_parameters_file="$(arg joint_limit_params)"
    kinematics_parameters_file="$(arg kinematics_params)"
    physical_parameters_file="$(arg physical_params)"
    visual_parameters_file="$(arg visual_params)"
    transmission_hw_interface="$(arg transmission_hw_interface)"
    safety_limits="$(arg safety_limits)"
    safety_pos_margin="$(arg safety_pos_margin)"
    safety_k_position="$(arg safety_k_position)"
  />

  <!-- Configure self collision properties per link -->
  <gazebo reference="${prefix}shoulder_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}upper_arm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}forearm_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_1_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_3_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}wrist_2_link">
    <selfCollide>true</selfCollide>
  </gazebo>
  <gazebo reference="${prefix}ee_link">
    <selfCollide>true</selfCollide>
  </gazebo>

  <!-- ############### Bota Systems SensONE FT SENSOR-->
  <xacro:include filename="$(find rokubimini_description)/urdf/BFT_SENS_ECAT_M8.urdf.xacro" />

  <xacro:BFT_SENS_ECAT_M8 prefix="${prefix}ft_sensor" />

  <joint name="ur_to_sensor" type="fixed">
    <parent link="${prefix}tool0" />
    <child link="${prefix}ft_sensor_mounting" />
    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0" />
  </joint>
  
  <!-- ############### GRIPPER ################ -->
  <xacro:include filename="$(find chargepal_description)/urdf/robotiq_2f_140/robotiq_arg2f_140_model_macro.xacro" />
  
  <!-- gripper coupling (ISO 9409-1-50-4-M / Robotiq AGC-CPL-062-002_20171219) -->
  <xacro:property name="coupling_radius" value="0.0375" />
  <xacro:property name="coupling_outer_length" value="0.014" />
  <xacro:property name="coupling_inner_length" value="0.008" />
  <xacro:property name="coupling_mass" value="0.100" />

  <joint name="${prefix}ft300_coupling_joint" type="fixed">
    <origin xyz="0 0 ${coupling_outer_length / 2}" rpy="0 0 ${-pi/2}" />
    <parent link="${prefix}ft_sensor_wrench" />
    <child link="${prefix}gripper_coupling_link" />
  </joint>

  <link name="${prefix}gripper_coupling_link">
    <origin xyz="0 0 0" rpy="0 0 0" />
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="${coupling_radius}" length="${coupling_outer_length}" />
      </geometry>
      <xacro:insert_block name="material_black" />
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder radius="${coupling_radius}" length="${coupling_outer_length}" />
      </geometry>
    </collision>
    <xacro:cylinder_inertial_ radius="${coupling_radius}" length="${coupling_outer_length}" mass="${coupling_mass}">
      <origin xyz="0.0 0.0 0.0" rpy="0 0 0" />
    </xacro:cylinder_inertial_>
  </link>
  <gazebo reference="${prefix}gripper_coupling_link">
    <material>Gazebo/FlatBlack</material>
  </gazebo>
  
  <joint name="${prefix}coupling_2f140_joint" type="fixed">
    <origin xyz="0 0 ${coupling_inner_length / 2}" rpy="0 0 0" />
    <parent link="${prefix}gripper_coupling_link" />
    <child link="${prefix}gripper_robotiq_arg2f_base_link" />
  </joint>
  <xacro:robotiq_arg2f_140 prefix="${prefix}gripper_"/>


  <!-- ################ GAZEBO ################ -->
  <!-- Common -->
  <xacro:include filename="$(find chargepal_description)/urdf/utils/common.gazebo.xacro" />
  <xacro:controller_plugin_gazebo robot_namespace=""/>

  <!-- Robotiq gripper -->
  <xacro:include filename="$(find chargepal_description)/urdf/robotiq_2f_140/robotiq_2f_140.gazebo.xacro" />
  <xacro:mimic_joint_plugins_gazebo prefix="gripper_" robot_namespace=""/>

</robot>